sql/sql_type.cc

```diff
@@ -7845,3 +7845,84 @@ Type_handler_time_common::create_literal_item(THD *thd,
   literal_warn(thd, item, str, length, cs, &st, "TIME", send_error);
   return item;
 }
+
+
+/** Checks for any blog type.
+@param[in]	type	MariaDB field type
+@return	true if type is BLOB and false otherwise */
+static bool is_any_blob_type(enum_field_types type)
+{
+  return type == MYSQL_TYPE_TINY_BLOB || type == MYSQL_TYPE_MEDIUM_BLOB ||
+         type == MYSQL_TYPE_LONG_BLOB || type == MYSQL_TYPE_BLOB;
+}
+
+static bool is_instant_alter_charset_type(enum_field_types type)
+{
+  return type == MYSQL_TYPE_VARCHAR || type == MYSQL_TYPE_STRING ||
+         is_any_blob_type(type);
+}
+
+extern "C" MY_CHARSET_HANDLER my_charset_utf8_handler;
+extern "C" MY_CHARSET_HANDLER my_charset_utf8mb4_handler;
+
+bool Type_handler::Can_convert_charset_instantly(
+    const Create_field &create_field)
+{
+  const Field &field= *create_field.field;
+
+  if (!field.stored_in_db())
+    return false;
+
+  const enum_field_types old_type= field.type_handler()->field_type();
+
+  if (!is_instant_alter_charset_type(old_type))
+  {
+    return false;
+  }
+
+  const enum_field_types new_type= create_field.type_handler()->field_type();
+
+  if (is_any_blob_type(new_type))
+  {
+    if (!is_any_blob_type(old_type))
+      return false;
+  }
+  else if (old_type != new_type)
+  {
+      return false;
+  }
+
+  if (field.char_length() != create_field.char_length)
+    return false;
+
+  const CHARSET_INFO *old_ci= field.charset();
+  const CHARSET_INFO *new_ci= create_field.charset;
+
+  const char * old_csname = old_ci->csname;
+  const char * new_csname = new_ci->csname;
+
+  if (strcmp(old_csname, new_csname) &&
+      !((!strcmp(old_csname, MY_UTF8MB3) && !strcmp(new_csname, MY_UTF8MB4)) ||
+        (!strcmp(old_csname, "ascii") && !strcmp(new_csname, MY_UTF8MB3)) ||
+        (!strcmp(old_csname, "ascii") && !strcmp(new_csname, MY_UTF8MB4)) ||
+        (!strcmp(old_csname, "ascii") && !strcmp(new_csname, "latin1")) ||
+        (!strcmp(old_csname, "ascii") && !strcmp(new_csname, "koi8")) ||
+        (!strcmp(old_csname, "ascii") && !strcmp(new_csname, "koi8u"))))
+  {
+    return false;
+  }
```

Comment:
Can you invoke `strcmp(old_csname, "ascii")` only once, and write a bunch of `(!strcmp(new_csname, …) || !strcmp(new_csname, …))`?

Are these really all the character sets that are proper supersets of `"ascii"`? Are `"latin2"` to `"latin7"` not proper supersets of ASCII? Also GBK, UJIS, SJIS, BIG5 would seem to be supersets of ASCII.

Do we support converting anything to `"binary"` (`my_charset_bin->number == 63`) without any validation? If we do, then we should also support that one instantly.

Please confirm your findings with @abarkov.

Reply:
> Can you invoke strcmp(old_csname, "ascii") only once, and write a bunch of (!strcmp(new_csname, …) || !strcmp(new_csname, …))?

> Are these really all the character sets that are proper supersets of "ascii"? Are "latin2" to "latin7" not proper supersets of ASCII? Also GBK, UJIS, SJIS, BIG5 would seem to be supersets of ASCII.

This is fixed and added.

> Do we support converting anything to "binary" (my_charset_bin->number == 63) without any validation? If we do, then we should also support that one instantly.

No, we don't support it.

Reply:
As a matter of fact, all character sets, with exception of really exotic SWE7, EBCDIC variants, and UCS/UTF<more-than-8>,  are superset of ASCII. pinging  @abarkov , who may want to correct this statement

Reply:
> Do we support converting anything to `"binary"` (`my_charset_bin->number == 63`) without any validation? If we do, then we should also support that one instantly.

Nope. It's disallowed by server code.

Reply:
Conversion to "binary" should be supported without validation.

Do you have an example script that demonstrates it's not supported?
You might hit some bug.

Reply:
Here is the test for that case https://github.com/MariaDB/server/blob/10.4/mysql-test/suite/innodb/t/instant_alter_charset.test#L143